internals/osutil/mkdirallchown.go

```diff
@@ -31,16 +31,23 @@ var mu sync.Mutex
 
 // MkdirAllChown is like os.MkdirAll but it calls os.Chown on any
 // directories it creates.
-func MkdirAllChown(path string, perm os.FileMode, uid sys.UserID, gid sys.GroupID) error {
+func MkdirAllChown(path string, perm os.FileMode, flags AtomicWriteFlags, uid sys.UserID, gid sys.GroupID) error {
 	mu.Lock()
 	defer mu.Unlock()
 	if uid == NoChown && gid == NoChown {
-		return os.MkdirAll(path, perm)
+		err := os.MkdirAll(path, perm)
+		if err != nil {
+			return err
+		}
+		if flags&AtomicWriteChmod != 0 {
+			return os.Chmod(path, perm)
```

Comment:
I think we have to be careful here: the doc comment says "it calls os.Chown on any directories it creates", so presumably the same should be true of chmod. However, the above code means it will only chmod the last/lowest directory it creates, not any parents that may have been created. So I think we need to change the logic to use the recursive helper function in both cases so we can chown and chmod every directory we create.